### DISCUSSION TITLE
VAD probability very low using ONNX on Android

### DISCUSSION THREAD

**Discussion by lhr0909** [snakers4/silero-vad#284] [Q&A] [answered]
Hi, I am building an Android App trying to use the ONNX VAD model to determine speech in the AudioRecord stream. I got the ONNX model integrated but the probability numbers are way off. I have to set the threshold to 0.15 to be more accurate.

I realize that the audio recording amplitude was a bit too low, so I multiply by a gain, but it doesn't seem to solve the issue.

[here](https://github.com/lhr0909/live-subtitles-rokid-ar/blob/f19ecc197d3bee6484fa7145f73607bb90d77869/app/src/main/java/chat/senses/livesubs/TranscribeViewModel.kt#L196) is the integration point and please let me know if there is anything I have done wrong. I can also provide a sample recording.

[voice-0e0d0a5c-4d51-4209-a51d-c0e32b39ede3.pcm.zip](https://github.com/snakers4/silero-vad/files/10370220/voice-0e0d0a5c-4d51-4209-a51d-c0e32b39ede3.pcm.zip)

The zip file contains a raw PCM mono recording, 16000Hz sample rate, big-endian float format. This audio file is already multiplied by a `10^(6/20)` gain.

**Comment by snakers4** [comment, marked as answer]
Hi,

I cannot really comment on the language used for Android, but looks like your code is not stateful. I.e. the VAD should keep the state during the session. With ONNX the best illustration is here:

https://github.com/snakers4/silero-vad/blob/e7c4539106bbfa14b3e15dee6bf761375a3c85f1/utils_vad.py#L63-L68

Looks like you always re-create these tensors:

https://github.com/lhr0909/live-subtitles-rokid-ar/blob/f19ecc197d3bee6484fa7145f73607bb90d77869/app/src/main/java/chat/senses/livesubs/TranscribeViewModel.kt#L199-L200

What it means in practical terms, on each chunk the VAD "thinks" that this is a new audio. Also note that there is quite a bit of logic in `get_speech_timestamps` which probably must be ported.

What I also recommend as debugging tool is to run you audio using the provided utils with `visualize_probs = True` (e.g. in Colab) to see what the internal states look like for your audio to set the thresholds properly.

**Reply by lhr0909** [reply]
Thank you for the quick response! @snakers4 after storing the state vectors it is working now!

**Reply by snakers4** [reply]
Also make sure to tune the hyper-params for thresholds properly.

**Reply by lhr0909** [reply]
@snakers4 I was using the torch hub model, and I didn't recall having to do the hyper-params tuning. Is it also needed for the PyTorch model?

**Reply by snakers4** [reply]
The `torch.hub` model, the PyTorch model and the ONNX model are essentially the same model.
`torch.hub` is just used as a very light packaging tool and the underlying models for `.jit` and `.onnx` are the same.
If you need to get the hyper-params once for you app it is worth doing:

https://github.com/snakers4/silero-vad/blob/e7c4539106bbfa14b3e15dee6bf761375a3c85f1/utils_vad.py#L161-L172

Typically it can be done quite quickly simply by looking at the probability chart.

**Comment by lhr0909** [comment]
For anyone that is trying to integrate the ONNX model into Android, this is a working version I put together in a ViewModel that takes a live stream of audio PCM input and returns the speech buffers. It is a bit rough since I don't have too much Kotlin experience, but it should show how the ONNX model is used in Kotlin/Java/Android.

I will spend some time to set up an Android Library just for Silero VAD integration and incorporate the `get_speech_timestamps` method from Python.

https://github.com/lhr0909/live-subtitles-rokid-ar/blob/c6c5dd34b53150b79c75b214e6c8facbc552d7ec/app/src/main/java/chat/senses/livesubs/TranscribeViewModel.kt
